### DISCUSSION TITLE
Cannot deploy the app from github using  Vercel REST API

### DISCUSSION THREAD

**Discussion by ghubnab99** [vercel/community#3403] [Help] [answered]
### Summary

We want to deploy an app from a connected github repo by invoking the Vercel Rest API: 
POST: Create a new deployment: (https://vercel.com/docs/rest-api/endpoints#create-a-new-deployment)
https://api.vercel.com/v13/deployments

We were able to successfully create a new project through the rest api for the connected gihub repo, but are not able to create a new deployment.

We are not sure, if we're not getting the concepts of the projects vs deployments correctly.

Can we create a project and then deployment simultaneously in one single call through any specific API Endpoint.

Furthermore, the API Reference also does not have any examples or detailed information on the API params

Kindly requesting if anyone could please guide us in the correct direction.

Thank you

### Steps to Reproduce

POST: https://api.vercel.com/v13/deployments
(Auth Bearer Token is already being attached to the POST Request)

NOTE that we want to deploy the app from a connect github repo

{
    "branch": {
      "project_id": "prj_ODylioBmMtTvWRk9Qvqt3ZvQKYKV",
      "name": "new-deployment",
      "current_state": "init",
      "primary": true
    },
    "name": "my-instant-deployment",
    "framework": "nextjs"
}

Response: 400 Bad Request

{
    "error": {
        "code": "bad_request",
        "message": "Invalid request: \"files\" field should be an array"
    }
}

**Comment by amyegan** [comment, marked as answer]
Hey @ghubnab99. It sounds like you're trying to deploy from an existing git commit. In that case, here is the body data I used to create a new deployment for one of my test sites.

```
{
  "gitSource": {
      "ref" : "main",
      "repoId": "668449998",
      "type": "github",
      "sha": "a59d548"
      },
  "name": "nextjs-with-supabase-test"
}
```

If `gitSource` is not provided, then a `files` array is expected. That is why you're seeing `"Invalid request: "files" field should be an array"`.

**Reply by ghubnab99** [reply]
Thanks a lot @amyegan!

We also had to add the block for "projectSettings" and it automatically created a project and deployed the app.

    "projectSettings": {
      "framework": "nextjs"
    },   

But what we want is that can we specify a custom domain / sub-domain that the project should be deployed with?

And we also noticed that in the Vercel dashboard, the project summary shows, "No Git repository Connected". What extra option do we need to pass in the create deployment api that it gets connected to the git repo.

I believe the connection to the git repo implies that any further commits to the git repo would auto-trigger a deployment.

Thanks

**Reply by amyegan** [reply]
To connect a GitHub repo when creating a new project, you need to include 
```
"gitRepository": {
  "repo": "SOME_STRING_VALUE",
  "type": "github"
},
```

If `gitRepository` is defined, any push to the connected Git repo will be automatically deployed. You can also connect a Git repo from the Git tab in Project Settings.

**Reply by ghubnab99** [reply]
Thanks a lot again @amyegan!
We'd really wish if this sort of explanation was also present in the API Reference Docs, it would be much easier.

One more question being, how can we specify a custom domain / sub-domain that the project should be deployed with?

After we connect our domain to our Vercel account and enable wildcards, can we specify, like a custom sub-domain in the deployment API params, to create the new deployment with the custom sub-domain.

Kindly guide us.
Thank you

**Reply by mondragon-ai** [reply]
@ghubnab99 

Just to ensure I am understanding your question correctly: 

Once you upload the domain to Vercel, you'd like to connect a custom domain/sub-domain, and redeploy the project with that custom domain? 

Here's what I am doing, that may help you:

STEP #1: Upload the domain
```
POST https://api.vercel.com/v10/projects/:projet_id/domains?teamId=teamID
{
    "name": "SUB_DOMAIN.DOMAIN.com" || "DOMAIN.com" ||
}
```
if you are connecting a regular (apex) domain you'd likely need to use the `76.76.21.21` in the A field of your domain host provider. Whereas as subdomain would require the CNAME credentials and value of cname.vercel-dns.com. 

STEP #2: Verify Domains
```
POST https://api.vercel.com/v9/projects/:projet_id/domains/:domain/verify?teamId=teamID
{
    "name": "SUB_DOMAIN.DOMAIN.com" || "DOMAIN.com" ||
}
```
Once you added the appropriate credentials to your DNS records, simply verify the domain. This may take a minute to reflect on your project. 

Otherwise, use the previous deployment ID and redeploy the project with the `v13/deployments` endpoint

Good luck!
